process/block/shardblock.go

```diff
@@ -1154,34 +1158,43 @@ func (sp *shardProcessor) createAndProcessCrossMiniBlocksDstMe(
 			continue
 		}
 
-		maxTxRemaining := uint32(maxTxInBlock) - nrTxAdded
-		currMBProcessed, currTxsAdded, hdrProcessFinished := sp.txCoordinator.CreateMbsAndProcessCrossShardTransactionsDstMe(hdr, maxTxRemaining, round, haveTime)
+		itemsAddedInBody := nrTxAdded
+		if itemsAddedInBody >= sp.blockSizeThrottler.MaxItemsToAdd() {
+			continue
+		}
 
-		// all txs processed, add to processed miniblocks
-		miniBlocks = append(miniBlocks, currMBProcessed...)
-		nrTxAdded = nrTxAdded + currTxsAdded
+		maxTxSpaceRemained := int32(sp.blockSizeThrottler.MaxItemsToAdd()) - int32(itemsAddedInBody)
+		maxMbSpaceRemained := int32(sp.blockSizeThrottler.MaxItemsToAdd()) - int32(itemsAddedInHeader) - 1
 
-		if currTxsAdded > 0 {
-			usedMetaHdrsHashes = append(usedMetaHdrsHashes, orderedMetaBlocks[i].hash)
-		}
+		if maxTxSpaceRemained > 0 && maxMbSpaceRemained > 0 {
+			currMBProcessed, currTxsAdded, hdrProcessFinished := sp.txCoordinator.CreateMbsAndProcessCrossShardTransactionsDstMe(
+				hdr,
+				uint32(maxTxSpaceRemained),
+				uint32(maxMbSpaceRemained),
+				round,
+				haveTime)
 
-		if !hdrProcessFinished {
-			break
-		}
+			// all txs processed, add to processed miniblocks
+			miniBlocks = append(miniBlocks, currMBProcessed...)
+			nrTxAdded = nrTxAdded + currTxsAdded
 
-		lastMetaHdr = hdr
-	}
+			if currTxsAdded > 0 {
+				usedMetaHdrsHashes = append(usedMetaHdrsHashes, orderedMetaBlocks[i].hash)
+			}
 
-	sp.mutUsedMetaHdrsHashes.Lock()
```

Comment:
put these lines back. I saw you moved out from the function, but it does not change the functionality at all.